### ISSUE TITLE
[Bug]: Custom React hook in a typescript class method trigger an unexpected error

### ISSUE THREAD

**Issue by nbrignol** [jsx-eslint/eslint-plugin-react#3564]
### Is there an existing issue for this?

- [X] I have searched the existing issues and my issue is unique
- [X] My issue appears in the command-line and not only in the text editor

### Description Overview

When I create a custom hook in a simple Typescript class, ESLINT is giving me this error : 

`React Hook "useState" cannot be called in a class component. React Hooks must be called in a React function component or a custom React Hook function  react-hooks/rules-of-hooks`

However, this is not an expected error because I am not working  with a class component.

IMHO It should not trigger any error OR at least trigger a specific, adapted error if anything.

**Standard implementation (classic function based) for reference :** 

```ts
function useMyCustomHookFunction(){
    return useState("this-is-a-test-1");
}
```

```ts
const [test1, setTest1] = useMyCustomHookFunction()

return (
    <>
         <p>From function: {test1}</p>
    </>
)
```

This works well, my output is : 
```
From function: this-is-a-test-1
```

**Class+method based implementation :**
This implementation trigger the unexpected error.

```ts
class TestHook {

    private initialValue = "this-is-a-test-2"
    
    public useMyCustomHookMethod(){
        return useState(this.initialValue);
    }

}

```

```ts
const testInstance = new TestHook()
const [test2, setTest2] = testInstance.useMyCustomHookMethod()

return (
    <>
        <p>From method: {test2}</p>
    </>
)
```
This works also well but trigger the error. My output is : 
```
From method: this-is-a-test-2
```

### Expected Behavior

I expect no warning or error when implementing a custom hook in a method.

### eslint-plugin-react version

7.32.2

### eslint version

8.36.0

### node version

v18.14.2

**Comment by JulesBlm** [comment]
First of all, this is not the right repository to report this bug `react-hooks/rules-of-hooks` is a warning from `eslint-plugin-react-hooks`, not `eslint-plugin-react`. `eslint-plugin-react-hooks` which is a package that lives in the [React monorepo](https://github.com/facebook/react/tree/main/packages/eslint-plugin-react-hooks).

Second, what you're doing here is very very unusual in React. There is no reason to put the `useState` hook in a class, React keeps track of the state value for you. I don't think this is an unexpected error, hooks are not meant to be called from classes. 

> [A]lways use Hooks at the top level of your **React function**, before any early returns
[Rules of Hooks](https://legacy.reactjs.org/docs/hooks-rules.html#only-call-hooks-at-the-top-level)

**Comment by nbrignol** [comment]
Thanks @JulesBlm for your response. I will fill a ticket in the right repository. I'll close this one.

And about the useState in a class method, as far as I know a class method is a function, so for my this is either : a false negative that should be removed; or an actual rule that need a specific message (that one talk about class _component_, which is not the case at all).

**Comment by ljharb** [comment]
Hooks can only be used in function _components_ - class methods are functions, but not components, so it's perfectly correct.
